### ISSUE TITLE
Changing parameters of active session

### ISSUE THREAD

**Issue by VividVisions** [openNDS/openNDS#245]
With RADIUS-based access control, one can often change parameters (speed, quota, etc…) of active sessions (or disconnect them) by sending CoA (Change of Authorization) packets from the auth server to the access point. I haven't found anything similar in the openNDS documentation. Did I overlook anything? Is something like this possible at all?

**Comment by bluewavenet** [comment]
@VividVisions 
Changing parameters of an active session is not currently supported. There has been little or no demand for this to warrant its support.
The only use for this that I can think of is to extend session duration or data quotas when additional payment is made - however paid captive portals are very much in the realm of legacy services. These days the majority of Captive Portals are to provide free Internet connections at public venues. Any venue demanding payment for connection is going to loose customers to competing venues that provide free access or be ignored and 4g mobile data used instead.
For this reason (amongst others) I would also class RADIUS based access control as a legacy service, support only being required in a large corporate environment where existing "legacy" RADIUS services exist.

> Did I overlook anything?

No

> Is something like this possible at all?

Yes, possible - but is development of it justified?

**Comment by VividVisions** [comment]
>  paid captive portals are very much in the realm of legacy services

Yes, I agree. 

The use case why I am asking this is access for contact information, which is pretty common. Users enter their e-mail address, get a couple of minutes of access to check their inbox for the confirmation e-mail where they have to click a link to gain full access.

**Comment by bluewavenet** [comment]
@VividVisions 

> which is pretty common

Yes it is.
Using openNDS, self account creation is commonly achieved by initially allowing the few minutes of access and clicking the email link deauthenticates them and sends them to a login page where they can only use the previously registered email address.
There is no need to update the active session. It is achieved purely with scripting and without any modifications to openNDS binaries.

**Comment by VividVisions** [comment]
I see. How would I achieve this without a login page but with automatic reauthentication when the link has been clicked? 
Is there a dynamic MAC address based authentication mechanism (controlled/configured by the FAS)?

**Comment by bluewavenet** [comment]
@VividVisions 
Re-authenticating automatically by clicking a link is potentially rather insecure. Sure, it is possible, but what if a second device is used to click on the link? Consider also that most new smart devices now use randomised mac addresses and can change mac between connections.
openNDS handles all this by forcing deauthentication and requiring the user to enter the previously created credentials.

**Comment by VividVisions** [comment]
@bluewavenet 

> Re-authenticating automatically by clicking a link is potentially rather insecure. 

Not if this link just adds the MAC address of the original device to a list of pre-authenticated devices. The question is, how to communicate this list between FAS and access point. And how to trigger the reconnect/re-authentication.

> […] randomised mac addresses and can change mac between connections.

To my knowledge both iOS/iPadOS and Android use the same randomised MAC address per network/SSID, which only changes if one deletes all settings on their device or if the SSID/network changes.

**Comment by bluewavenet** [comment]
@VividVisions 
Re: Random mac - that was my understanding too, but in practice I see it changing if the user tells the device to "forget" the ssid then reconnects - sometimes necessary when the device CPD gets into a deadlock.

Having said all this, I do see the place for a re-auth mode where timeouts and quotas are updated, particularly with the upcoming new rfc8910/8908 CPI implementation - I will look closely at the detail....  I think it will be easy to implement.

**Comment by bluewavenet** [comment]
@VividVisions 

> Sounds interesting!

Indeed. It opens a lot of possibilities. One being that it will make IPv6 support easy to achieve, not that IPv6 support is needed yet.....
The only issue is that very few client devices support it yet. A small minority of Android 11 devices are rumoured to support it, but I have not found one yet. I don't have access to any Apple devices at the moment.
We will probably have to roll out to a live venue to log devices that try CPI first... ;-)
